edgelet/provisioning/src/provisioning.rs

```diff
@@ -128,12 +192,96 @@ impl Provision for ManualProvisioning {
                 hub_name: hub,
                 reconfigure: ReprovisioningStatus::DeviceDataNotUpdated,
                 sha256_thumbprint: None,
+                credentials: None,
             })
             .map_err(|err| Error::from(err.context(ErrorKind::Provision)));
         Box::new(result.into_future())
     }
 }
 
+pub struct ExternalProvisioning<T, U> {
+    client: T,
+
+    // ExternalProvisioning is not restricted to a single HSM implementation, so it uses
+    // PhantomData to be generic on them.
+    phantom: PhantomData<U>,
+}
+
+impl<T, U> ExternalProvisioning<T, U> {
+    pub fn new(client: T) -> Self {
+        ExternalProvisioning {
+            client,
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<T, U> Provision for ExternalProvisioning<T, U>
+where
+    T: 'static + ExternalProvisioningInterface,
+    U: 'static + Activate + KeyStore + Send,
+{
+    type Hsm = U;
+
+    fn provision(
+        self,
+        mut _key_activator: Self::Hsm,
+    ) -> Box<dyn Future<Item = ProvisioningResult, Error = Error> + Send> {
+        let result = self
+            .client
+            .get_device_provisioning_information()
+            .map_err(|err| Error::from(err.context(ErrorKind::Provision)))
+            .and_then(move |device_provisioning_info| {
+                info!(
+                    "External device registration information: Device \"{}\" in hub \"{}\"",
+                    device_provisioning_info.device_id(),
+                    device_provisioning_info.hub_name()
+                );
+
+                let credentials_info = device_provisioning_info.credentials();
+                let credentials = if let "symmetric-key" = credentials_info.auth_type() {
+                        match credentials_info.source() {
+                            "payload" => {
+                                credentials_info.key().map_or_else(
+                                    || {
+                                        info!(
+                                            "A key is expected in the response with the 'symmetric-key' authentication type and the 'source' set to 'payload'.");
+                                        Err(Error::from(ErrorKind::Provision))
+                                    },
+                                    |key| {
+                                        Ok(Credentials { auth_type: AuthType::SymmetricKey(SymmetricKeyCredential{ key: Some(key.to_string()) }), source: CredentialSource::Payload })
```

Comment:
Break this up into multiple lines please. (rustfmt doesn't work reliably at deep nesting.)

Reply:
Sure. Yeah, kind of expected rustfmt to take care of this.